src/coloring.rs

```diff
@@ -59,3 +67,209 @@ pub fn graph_greedy_color(py: Python, graph: &graph::PyGraph) -> PyResult<PyObje
     }
     Ok(out_dict.into())
 }
+
+struct MisraGriesAlgorithm<'a> {
+    graph: &'a StablePyGraph<Undirected>,
+    colors: DictMap<EdgeIndex, usize>,
+}
+
+impl<'a> MisraGriesAlgorithm<'a> {
+    pub fn new(graph: &'a StablePyGraph<Undirected>) -> Self {
+        let colors: DictMap<EdgeIndex, usize> = DictMap::new();
+        MisraGriesAlgorithm { graph, colors }
+    }
+
+    // Computes colors used at node u
+    fn get_used_colors(&self, u: NodeIndex) -> Vec<usize> {
```

Comment:
I think you'd want to use a `HashSet` for this since `Vec<T>.contains()` will have an `O(n)` lookup.